tests/rptest/clients/ping_pong.py

```diff
@@ -0,0 +1,197 @@
+            "socket.timeout.ms": 10000,
+            "connections.max.idle.ms": 0,
+            "reconnect.backoff.ms": 100,
+            "reconnect.backoff.max.ms": 10000,
+            "statistics.interval.ms": 0,
+            "api.version.request.timeout.ms": 10000,
+            "api.version.fallback.ms": 0,
+            "enable.auto.commit": False,
+            "auto.offset.reset": "earliest",
+            "isolation.level": "read_committed",
+            "session.timeout.ms": 10000,
+            "heartbeat.interval.ms": 300,
+            "fetch.wait.max.ms": 500
+        })
+        self.consumer.assign(
+            [TopicPartition(topic, partition, OFFSET_BEGINNING)])
+        self.stream = self.stream_gen()
+
+    def stream_gen(self):
+        while True:
+            msgs = self.consumer.consume(timeout=10)
+            for msg in msgs:
+                yield msg
+
+    def read_until(self, check, timeout_s):
+        begin = time.time()
+        while True:
+            if time.time() - begin > timeout_s:
+                raise KafkaException(KafkaError(KafkaError._TIMED_OUT))
+            for msg in self.stream:
+                offset = msg.offset()
+                value = msg.value().decode('utf-8')
+                key = msg.key().decode('utf-8')
+                if check(offset, key, value):
+                    return
+
+
+def expect(offset, key, value):
+    def check(ro, rk, rv):
+        if ro < offset:
+            return False
+        if ro == offset:
+            if rk != key:
+                raise RuntimeError(f"expected key='{key}' got '{rk}'")
+            if rv != value:
+                raise RuntimeError(f"expected value='{value}' got '{rv}'")
+            return True
+        raise RuntimeError(f"read offset={ro} but skipped {offset}")
+
+    return check
+
+
+class PingPong:
+    @staticmethod
+    def cold(brokers, topic, partition, logger):
+        bootstrap = ",".join(brokers)
+        consumer = LogReader(bootstrap)
+        consumer.init(group="ping_ponger1", topic=topic, partition=partition)
+        producer = SyncProducer(bootstrap)
+        producer.init()
+        return PingPong(consumer, producer, topic, partition, logger)
+
+    @staticmethod
+    def warmed(brokers, topic, partition, logger, timeout_s):
+        """
+        confluent_kafka isn't intelligent in finding the right broker
+        for issuing init_producer_id, by protocol the client asks a
+        random node and then the node proxy the request to the right
+        leader. then the node happen to be isolated the request hangs.
+        the protocol lacks the error code to ask client to try another
+        node.
+
+        the client sends init_producer_id on first produce so the only
+        way to make sure that the init_producer_id has executed is to
+        keep invoking produce until it succeeds
+        """
+        bootstrap = None
+
+        begin = time.time()
+        while True:
```

Comment:
This setup phase worries me.  The idea of the test is to check that a normal client will get prompt service from the cluster when a node is isolated or down -- are we saying that our customers' applications need to do this kind of initialization loop in order to avoid produce errors while a node is isolated or down?

Reply:
This is what we deal with. I've looked into source of franz-go and java client and return [the right error code](https://github.com/redpanda-data/redpanda/pull/4199/commits/e967424bde35529c632a52c20e6e62ebe232ec9a) on which they pick a random broker (least loaded broker) when current broker can't allocate producer id.

With confluent_kafka (librdkafka) when a randomly chosen broker isn't able to allocate id it keep retry on the same broker (at least this is the situation I saw in the logs). So the only option is to close the client, create a new producer and retry until it picks non-isolated node.

Reply:
Ok, so that puts us in an awkward position: switching the test to use a buggy client is not great, but I understand the motivation (it's easier to write this into python than to create an external Go program to do the ping-ponging).

I notice that we are not using the most recent version of confluent_kafka in setup.py, so maybe they already fixed this upstream?  If not, then we should follow up on the issue: our customers will hit this sooner or later

Reply:
Replaced weird init with retry on first produce (added retry argument). Fired an issue to `confluent_kafka` https://github.com/confluentinc/confluent-kafka-python/issues/1335